core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -200,14 +229,32 @@ class LogSegment(val log: FileMessageSet,
     val truncated = log.sizeInBytes - validBytes
     log.truncateTo(validBytes)
     index.trimToValidSize()
+    // A normally closed segment always appends the biggest timestamp ever seen into log segment, we do this as well.
+    timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp)
+    timeIndex.trimToValidSize()
     truncated
   }
 
+  def loadTimestamps(readToLogEnd: Boolean = false) {
+    // Get the last time index entry. If the time index is empty, it will return (-1, baseOffset)
+    if (log.iterator.hasNext)
+      rollingBaseTimestamp = log.iterator.next.message.timestamp
```

Comment:
Not sure I fully understand this. It seems we need to read the first message from the file anyway, whether at the startup time or when the first message is appended after startup. If the first message received after startup and the first message in the segment resides in the same disk block, we only page in once. Otherwise we still need to page in two blocks. So it seems the disk I/Os are the same? Do you mean we can save the disk I/O if there is no message appended to a partition after startup?